File: features/step_definitions/triangles.rb
Given("t ← triangle\\({point}, {point}, {point})") do |p1, p2, p3|
  @t = Rayz::Triangle.new(p1: p1, p2: p2, p3: p3)
end

Then("t.p1 = p1") do
  assert_equal(@t.p1, @p1)
end

Then("t.p2 = p2") do
  assert_equal(@t.p2, @p2)
end

Then("t.p3 = p3") do
  assert_equal(@t.p3, @p3)
end

Then("t.e1 = vector\\({float}, {float}, {float})") do |x, y, z|
  expected = Rayz::Vector.new(x: x, y: y, z: z)
  assert_equal(@t.e1, expected)
end

Then("t.e2 = vector\\({float}, {float}, {float})") do |x, y, z|
  expected = Rayz::Vector.new(x: x, y: y, z: z)
  assert_equal(@t.e2, expected)
end

Then("t.normal = vector\\({float}, {float}, {float})") do |x, y, z|
  expected = Rayz::Vector.new(x: x, y: y, z: z)
  assert_equal(@t.normal, expected)
end

When("xs ← local_intersect\\(t, r)") do
  @xs = @t.local_intersect(@r)
end

When(/^(n[123]) ← local_normal_at\(t, (point\(.+\))\)$/) do |var_name, point_str|
  # Parse the point string
  if point_str =~ /point\(([^,]+),\s*([^,]+),\s*([^)]+)\)/
    point = Rayz::Point.new(x: $1.to_f, y: $2.to_f, z: $3.to_f)
    normal = @t.local_normal_at(point)
    instance_variable_set("@#{var_name}", normal)
  end
end

Then("n1 = t.normal") do
  assert_equal(@n1, @t.normal)
end

Then("n2 = t.normal") do
  assert_equal(@n2, @t.normal)
end

Then("n3 = t.normal") do
  assert_equal(@n3, @t.normal)
end

File: python/features/steps/triangles.py
import pytest
from behave import given, then, use_step_matcher

from rayz.math_parser import parse_math
from rayz.triangle import Triangle
from rayz.tuple import Point, Vector

use_step_matcher("re")

_V = r"([A-Za-z][A-Za-z0-9_]*)"
_A = r"([^\s,)]+)"


@given(rf"{_V} ← triangle\({_V},\s*{_V},\s*{_V}\)")
def step_given_triangle(context, var, p1, p2, p3):
    setattr(context, var, Triangle(getattr(context, p1), getattr(context, p2), getattr(context, p3)))


_P = r"point\(([^\)]+)\)"


def _parse_point(spec: str) -> Point:
    x, y, z = [parse_math(v.strip()) for v in spec.split(",")]
    return Point(x, y, z)


@given(rf"{_V} ← triangle\({_P},\s*{_P},\s*{_P}\)")
def step_given_triangle_inline(context, var, p1, p2, p3):
    setattr(context, var, Triangle(_parse_point(p1), _parse_point(p2), _parse_point(p3)))


@then(rf"{_V}\.p1 = {_V}")
def step_then_triangle_p1(context, shape, pt):
    assert getattr(context, shape).p1 == getattr(context, pt)


@then(rf"{_V}\.p2 = {_V}")
def step_then_triangle_p2(context, shape, pt):
    assert getattr(context, shape).p2 == getattr(context, pt)


@then(rf"{_V}\.p3 = {_V}")
def step_then_triangle_p3(context, shape, pt):
    assert getattr(context, shape).p3 == getattr(context, pt)


@then(rf"{_V}\.e1 = vector\({_A},\s*{_A},\s*{_A}\)")
def step_then_triangle_e1(context, var, x, y, z):
    expected = Vector(parse_math(x), parse_math(y), parse_math(z))
    assert getattr(context, var).e1 == expected


@then(rf"{_V}\.e2 = vector\({_A},\s*{_A},\s*{_A}\)")
def step_then_triangle_e2(context, var, x, y, z):
    expected = Vector(parse_math(x), parse_math(y), parse_math(z))
    assert getattr(context, var).e2 == expected


@then(rf"{_V}\.normal = vector\({_A},\s*{_A},\s*{_A}\)")
def step_then_triangle_normal(context, var, x, y, z):
    expected = Vector(parse_math(x), parse_math(y), parse_math(z))
    actual = getattr(context, var).normal
    assert actual.x == pytest.approx(expected.x, abs=1e-5)
    assert actual.y == pytest.approx(expected.y, abs=1e-5)
    assert actual.z == pytest.approx(expected.z, abs=1e-5)


@then(rf"{_V} = {_V}\.normal")
def step_then_var_eq_shape_normal(context, result_var, shape_var):
    assert getattr(context, result_var) == getattr(context, shape_var).normal
